### DISCUSSION TITLE
Getting 403 errors from Google Apps Script

### DISCUSSION THREAD

**Discussion by rjwalters** [weather-gov/api#435] [Q&A] [answered]
I had a script that was working fine for several weeks, polling data from my local weather station once every 10 minutes. I am now getting 403 errors, similar to the open discussion @ #238 

Since the weather.gov api docs suggested that my `User-Agent` header could be to blame, I tried modifying it to include my email address -- but I quickly ran into a limitation of Google's Apps Scripts:

https://stackoverflow.com/questions/15067462/urlfetch-with-custom-user-agent-string

It seems there is no way to set a custom value for `User-Agent` for `UrlFetchApp`. I added an upvote on the open feature request page linked from stack overflow, but Google has not made any progress on the issue for several years so it seems unlikely to change.

Any suitable workarounds or other ideas for causes?

Alternatively, do people have substitute weather API recommendations, paid or free?

[edit] 

Confirmed using "myuseragent.net" that my custom `User-Agent` is indeed stripped from the `params` object:

```javascript
    // test UrlFetchApp's User-Agent override
    Logger.log(UrlFetchApp.fetch('http://www.myuseragent.net/', params).getContentText());
    const response = await UrlFetchApp.fetch(url, params);
    Logger.log(response.getContentText());
```

results in:

> User-Agent is<br /></h1><b>Mozilla/5.0 (compatible; Google-Apps-Script; beanserver; +https://script.google.com; id: UAEm...TkwOkwU)</b>

and the api.weather.gov response:

```html
<HTML><HEAD>
<TITLE>Access Denied</TITLE>
</HEAD><BODY>
<H1>Access Denied</H1>
You don't have permission to access "http://api.weather.gov/stations/K***/observations/latest" on this server.
<P>Reference #18.4fec3817.1626985111.4b1ca6b1</P>
</BODY>
</HTML>
```

**Comment by scadergit** [comment, marked as answer]
Thanks for providing the reference id. It took me a moment, but I tracked that down to, "Detected request from anonymous proxy	." The development team does not manage the web application firewall policies, but I'm guessing this request goes through a Google service proxy? It doesn't matter what the user agent string is, as long as one is provided. A custom string is only helpful to provide a contact in the event we have to take action due to abuse, which is very rare.

**Reply by scadergit** [reply]
If you are using a VPN proxy service it might be related, see https://blogs.akamai.com/2019/03/prevent-access-to-unauthorized-viewers-with-enhanced-proxy-detection.html

**Reply by rjwalters** [reply]
That's helpful -- I do usually work with a VPN turned on (goldenfrog's "vyprvpn") so maybe that was the actual issue rather than `User-Agent`? 

** It seems to be working with the VPN turned off! **

**Reply by scadergit** [reply]
Well I learned something today. =)
